tests/repository/test_upload.py

```diff
@@ -0,0 +1,118 @@
+import logging
+import pytest
+import json
+import io
+
+from pubtools.pulplib import (
+    Repository,
+    FileRepository,
+    Task,
+    DetachedException,
+    TaskFailedException,
+)
+
+
+def test_upload_detached():
+    """upload_file raises if called on a detached repo"""
+    with pytest.raises(DetachedException):
+        FileRepository(id="some-repo").upload_file("some-file")
+
+
+def test_upload_file(client, requests_mocker, tmpdir, caplog):
+    """test upload a file to a repo in pulp"""
+
+    logging.getLogger().setLevel(logging.INFO)
+    caplog.set_level(logging.INFO)
+
+    repo_id = "repo1"
+    repo = FileRepository(id=repo_id)
+    repo.__dict__["_client"] = client
+
+    client._CHUNK_SIZE = 20
+
+    request_body = {
+        "_href": "/pulp/api/v2/content/uploads/cfb1fed0-752b-439e-aa68-fba68eababa3/",
+        "upload_id": "cfb1fed0-752b-439e-aa68-fba68eababa3",
+    }
+    upload_id = request_body["upload_id"]
+
+    import_report = {
+        "result": {},
+        "error": {},
+        "spawned_tasks": [{"_href": "/pulp/api/v2/tasks/task1/", "task_id": "task1"}],
+    }
+
+    tasks_report = [{"task_id": "task1", "state": "finished"}]
+
+    somefile = tmpdir.join("some-file.txt")
+    somefile.write(b"there is some binary data:\x00\x01\x02")
+
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/content/uploads/", json=request_body
+    )
+    requests_mocker.put(
+        "https://pulp.example.com/pulp/api/v2/contents/%s/0" % upload_id, json=[]
+    )
+    requests_mocker.put(
+        "https://pulp.example.com/pulp/api/v2/contents/%s/20" % upload_id, json=[]
+    )
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/repositories/%s/actions/import_upload/"
+        % repo_id,
+        json=import_report,
+    )
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/tasks/search/", json=tasks_report
+    )
+    requests_mocker.delete(
+        "https://pulp.example.com/pulp/api/v2/content/uploads/%s/" % upload_id, json=[]
+    )
+
+    assert repo.upload_file(str(somefile)).result() == [
+        Task(id="task1", succeeded=True, completed=True)
+    ]
+
+    # it' possible the delete has been called while doing assertion
+    assert requests_mocker.call_count == 5 or 6
```

Comment:
This doesn't make sense.  Please consider the precedence of each part of this expression.

```
assert requests_mocker.call_count == 5 or 6
```

Is like writing:

```
assert( (requests_mocker.call_count == 5) or 6 )
```

And since 6 is always true, the assertion will always pass.

Try simplified version and note this assertion always passes even though the == part obviously is always false:

```
>>> assert 1 == 2 or 6
```

I think what you wanted was:

```
requests_mocker.call_count in [5, 6]
```

But actually, I think maybe you should rewrite this to wait for the deletion anyway during the test.  I mean, for example:

```
# Upload deletion is async and detached from our future - wait until it's done
for _ in range(0, 100):
  if requests_mocker.call_count == 6:
    break
  time.sleep(0.01)
assert requests_mocker.call_count == 6
```

And then you can actually verify in request_history that the deletion happened.

Reply:
ah, overlooked this. 
Yeah, I thought about using time.sleep once, but just felt a little strange to sleep in the test.